diff --git a/opcodes/hexagon-dis.c b/opcodes/hexagon-dis.c
index 73e0bf6..9e4fa22 100644
--- a/opcodes/hexagon-dis.c
+++ b/opcodes/hexagon-dis.c
@@ -79,19 +79,6 @@ hexagon_decode_inst(
   int bytes;
   char *str;
 
-  /* Decide if we have a 16-bit instruction */
-  status = (*info->read_memory_func)(address, buffer, 2, info);
-  if (status != 0) {
-    (*info->memory_error_func)(status, address, info);
-    return -1;
-  }
-  if (info->endian == BFD_ENDIAN_LITTLE) {
-    insn = bfd_getl16(buffer);
-  }
-  else {
-    insn = bfd_getb16(buffer);
-  }
-
   status = (*info->read_memory_func)(address, buffer, 4, info);
   if (status != 0) {
     (*info->memory_error_func)(status, address, info);
@@ -107,14 +94,14 @@ hexagon_decode_inst(
   /* disassemble  */
   bytes = hexagon_dis_inst(address, insn, instrBuffer, info);
 
-  /* display the disassembly instruction  */
-  if (bytes == 2) {
-    (*func) (stream, "    %04x ", insn);
-  }
-  else {
-    (*func) (stream, "%08x ", insn);
-  }
-  (*func) (stream, "    ");
+///* display the disassembly instruction  */
+//if (bytes == 2) {
+//  (*func) (stream, "    %04x ", insn);
+//}
+//else {
+//  (*func) (stream, "%08x ", insn);
+//}
+//(*func) (stream, "    ");
 
   /* Print the instruction buffer
      Watch out for placeholders where we want
diff --git a/opcodes/hexagon-opc.c b/opcodes/hexagon-opc.c
index a49f709..a115143 100644
--- a/opcodes/hexagon-opc.c
+++ b/opcodes/hexagon-opc.c
@@ -994,13 +994,16 @@ size_t hexagon_opcodes_count;
   (char *i, size_t n, const hexagon_operand_arg args [])
 #define MAP_FUNCTION(TAG) HEXAGON_MAP_SIGN (TAG); \
   HEXAGON_MAP_SIGN (TAG)
+
+#define HEXAGON_MAP_FPTR(TAG) ((void*)&HEXAGON_MAP_FNAME(TAG))
+
 #define DEST (i), (n)
 #define GET_OP_VAL(N) ((int) args [N].value)
 #define GET_OP_STR(N) (args [N].string [0] == '#' \
                        ? args [N].string + 1 \
                        : args [N].string)
-
-#define sprintf snprintf
+#undef sprintf
+#define sprintf(buf, len, fmt, ...) snprintf(buf, len, fmt, ##__VA_ARGS__)
 #  include "opcode/hexagon_asm_maps.h"
 #undef sprintf
 
@@ -2009,7 +2012,9 @@ hexagon_reg_num
                 continue;
 
               *input = regs + len - (*name? 1: 0);
-              return ((int) (aliasn? *aliasn = i: i), (int) regn);
+              if (aliasn)
+                  *aliasn = i;
+              return  (int)regn;
             }
         }
     }
